packages/rich-text/src/component/use-selection-change-compat.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * WordPress dependencies
+ */
+import { useRefEffect } from '@wordpress/compose';
+
+/**
+ * Sometimes some browsers are not firing a `selectionchange` event when
+ * changing the selection by mouse or keyboard. This hook makes sure that, if we
+ * detect no `selectionchange` or `input` event between the up and down events,
+ * we fire a `selectionchange` event.
+ *
+ * @return {import('@wordpress/compose').RefEffect} A ref effect attaching the
+ *                                                  listeners.
+ */
+export function useSelectionChangeCompat() {
+	return useRefEffect( ( element ) => {
+		const { ownerDocument } = element;
+
+		function onDown( event ) {
+			const type = event.type === 'keydown' ? 'keyup' : 'pointerup';
+
+			function onCancel() {
+				ownerDocument.removeEventListener( type, onUp );
+				ownerDocument.removeEventListener(
+					'selectionchange',
+					onCancel
+				);
+				ownerDocument.removeEventListener( 'input', onCancel );
+			}
+
+			function onUp() {
+				onCancel();
+				ownerDocument.dispatchEvent( new Event( 'selectionchange' ) );
```

Comment:
What happens when the "selection" doesn't change, should we guard against it?

Reply:
Updated it. While I don't see an improvement in performance, we're still reducing the amount of times `createRecord` is called and I like this isolated polyfill approach making it clearer what is happening.